doc/specs/#1502 - Advanced Tab Switcher/spec.md

```diff
@@ -0,0 +1,219 @@
+
+```
+1 Git Moo
+2 shoo
+3 /c/
+4 /d
+5 /e/
+6 /f/
+7 /g/
+8 /h/
+9 /i/
+  /j/
+  /k/
+  /l/ (highlighted)
+```
+
+`Git Moo` is now associated with number 1, and the `foo`, `boo`, `Windows boofoo`, and `/c/Users/booboo` tabs are no longer visible.
+
+### Using the Switcher
+
+#### Opening the Tab Switcher
+
+The user will press a keybinding named `OpenTabSwitcher` to bring up the UI.
+The user will be able to change it to whatever they like.
+There will also be an optional `anchor` arg that may be provided to this keybinding.
+
+#### Keeping it open
+
+We use the term `anchor` to illustrate the idea that the UI stays visible as long as something is "anchoring" it down.
+
+So, when the `OpenTabSwitcher` keybinding is given the `anchor` arg, the first key of the keybinding will act as the `anchor` key. Holding that`anchor` key down will keep the switcher visible, and once the `anchor` key is released, the switcher will dismiss.
+
+If `OpenTabSwitcher` is not given the `anchor` arg, the switcher will stay visible even after the release of the key/keychord.
+
+#### Switching through Tabs
+
+The user will be able to navigate through the switcher with the following keybindings:
+
+- Switching Down: <kbd>tab</kbd> or <kbd>downArrow</kbd>
+- Switching Up: <kbd>shift+tab</kbd> or <kbd>upArrow</kbd>
+
+As the user is cycling through the tab list, the selected tab will be highlighted but the terminal won't actually switch focus to the selected tab.
+
+#### Closing the Switcher and Bringing a Tab into Focus
+
+There are two _dismissal_ keybinds:
+
+1. <kbd>enter</kbd> : brings the currently selected tab into focus and dismisses the UI.
+2. <kbd>esc</kbd> : dismisses the UI without changing tab focus.
+
+The following are ways a user can dismiss the UI, _whether or not_ the `Anchor` arg is provided to `OpenTabSwitcher`.
+
+1. The user can press a number associated with a tab to instantly switch to the tab and dismiss the switcher.
+2. The user can click on a tab to instantly switch to the tab and dismiss the switcher.
+3. The user can click outside of the UI to dismiss the switcher without bringing the selected tab into focus.
+4. The user can press any of the dismissal keybinds.
+
+If the `Anchor` arg is provided, then in addition to the above methods, the UI will dismiss upon the release of the `Anchor` key.
+
+Pressing the `OpenTabSwitcher` keychord again will _not_ close the Switcher.
+
+### Most Recently Used Order
+
+We'll provide a setting that will allow the list of tabs to be presented in either _in-order_ (how the tabs are ordered on the tab bar), or _Most Recently Used Order_ (MRU). MRU means that the tab that the terminal most recently visited will be on the top of the list, and the tab that the terminal has not visited for the longest time will be on the bottom.

```

Comment:
That's true, having the setting live in `OpenTabSwitcher` makes sense to give the user the ability to have a binding for both orders. I personally wanted to limit the user to just one order at a time, just for simplicity's sake. However, it makes sense to give that decision to the user instead of straight-up not allowing it.

Although, there is one scenario I'd like to flesh out in regards to having two bindings simultaneously. Lets say the keybinding for TabSwitcher in-order is <kbd>ctrl+tab</kbd> and TabSwitcher MRU is <kbd>ctrl+w</kbd>. If the user has `anchor` set to <kbd>ctrl</kbd>, what happens when the user holds <kbd>ctrl</kbd> and alternates between <kbd>tab</kbd> and <kbd>w</kbd>? Should the list of tabs just magically update to the correct order without dismissing the UI? Maybe the current open UI should just dismiss, then the UI would reappear with the correct order?

Reply:
Totally concur with Carlos on this. The TabSwitcher will eat all the keypresses w/o going through the `AppKeyBindings` while it's open, so I don't think we'll need to worry about <kbd>Ctrl+w</kbd>, <kbd>Ctrl+tab</kbd> invoking the MRU switcher then switching to in-order.